File: app/src/main/java/me/devsaki/hentoid/json/sources/LusciousGalleryMetadata.java
package me.devsaki.hentoid.json.sources;

import java.util.ArrayList;
import java.util.List;

import me.devsaki.hentoid.database.domains.ImageFile;
import me.devsaki.hentoid.enums.StatusContent;

@SuppressWarnings("unused, MismatchedQueryAndUpdateOfCollection")
public class LusciousGalleryMetadata {
    private PictureData data;

    private static class PictureData {
        private PictureInfoContainer picture;
    }

    private static class PictureInfoContainer {
        private PictureInfo list;
    }

    private static class PictureInfo {
        private PictureContainerMetadata info;
        private List<PictureMetadata> items;
    }

    private static class PictureContainerMetadata {
        private int total_pages;
    }

    private static class PictureMetadata {
        private String url_to_original;
    }

    public int getNbPages() {
        return data.picture.list.info.total_pages;
    }

    public List<ImageFile> toImageFileList() {
        return toImageFileList(0);
    }

    public List<ImageFile> toImageFileList(int offset) {
        List<ImageFile> result = new ArrayList<>();

        int order = offset;
        List<PictureMetadata> imageList = data.picture.list.items;
        for (PictureMetadata pm : imageList)
            result.add(new ImageFile(++order, pm.url_to_original, StatusContent.SAVED, imageList.size()));

        return result;
    }
}

File: app/src/main/java/me/devsaki/hentoid/json/sources/LusciousGalleryMetadata.kt
package me.devsaki.hentoid.json.sources

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass
import me.devsaki.hentoid.database.domains.ImageFile
import me.devsaki.hentoid.enums.StatusContent


@JsonClass(generateAdapter = true)
data class LusciousGalleryMetadata(
    val data: PictureData
) {
    @JsonClass(generateAdapter = true)
    data class PictureData(
        val picture: PictureInfoContainer
    )

    @JsonClass(generateAdapter = true)
    data class PictureInfoContainer(
        val list: PictureInfo
    )

    @JsonClass(generateAdapter = true)
    data class PictureInfo(
        val info: PictureContainerMetadata,
        val items: List<PictureMetadata>
    )

    @JsonClass(generateAdapter = true)
    data class PictureContainerMetadata(
        @Json(name = "total_pages")
        var totalPages: Int = 0
    )

    @JsonClass(generateAdapter = true)
    data class PictureMetadata(
        @Json(name = "url_to_original")
        val urlToOriginal: String
    )

    fun getNbPages(): Int {
        return data.picture.list.info.totalPages
    }

    fun toImageFileList(offset: Int = 0): List<ImageFile> {
        val result: MutableList<ImageFile> = ArrayList()
        var order = offset
        val imageList: List<PictureMetadata> = data.picture.list.items
        imageList.forEach {
            result.add(
                ImageFile.fromImageUrl(
                    ++order,
                    it.urlToOriginal,
                    StatusContent.SAVED,
                    imageList.size
                )
            )
        }
        return result
    }
}
